ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/services/SandboxCommandCompletionService.scala

```diff
@@ -53,106 +56,83 @@ class SandboxCommandCompletionService private (
       "Received request for completion subscription {}: {}",
       subscriptionId: Any,
       request)
-    val offsetOrError =
-      request.offset.fold[Future[Option[Ref.LedgerString]]](Future.successful(None))(
-        o =>
-          LedgerOffsetValidator
-            .validate(o, "offset")
-            .fold(
-              Future.failed, {
-                case domain.LedgerOffset.Absolute(value) => Future.successful(Some(value))
-                case domain.LedgerOffset.LedgerBegin => Future.successful(None)
-                case domain.LedgerOffset.LedgerEnd =>
-                  ledgerBackend.getCurrentLedgerEnd.map(Some.apply)
-              }
-          ))
-
-    Source
-      .fromFuture(offsetOrError)
-      .flatMapConcat(completionSourceWithOffset(request, _, subscriptionId))
+
+    val offsetOrError = for {
+      offset <- FieldValidations.requirePresence(request.offset, "offset")
+      convertedOffset <- LedgerOffsetValidator.validate(offset, "offset")
+    } yield convertedOffset
+
+    offsetOrError.fold(
+      Source.failed,
+      offset => completionSourceWithOffset(request, offset, subscriptionId))
   }
 
   private def completionSourceWithOffset(
       request: CompletionStreamRequest,
-      requestedOffset: Option[LedgerSyncOffset],
+      requestedOffset: domain.LedgerOffset,
       subscriptionId: String): Source[CompletionStreamResponse, NotUsed] = {
-    val requestingParties = request.parties.toSet
-    val requestedApplicationId = request.applicationId
-    ledgerBackend
-      .ledgerSyncEvents(requestedOffset)
-      .map { syncEvent =>
-        val checkpoint =
-          Some(
-            Checkpoint(
-              Some(fromInstant(syncEvent.recordTime)),
-              Some(LedgerOffset(LedgerOffset.Value.Absolute(syncEvent.offset)))))
-
-        syncEvent match {
-          case tx: LedgerSyncEvent.AcceptedTransaction
-              if isRequested(
-                requestedApplicationId,
-                requestingParties,
-                tx.applicationId,
-                tx.submitter) =>
+    //TODO: put these into a proper validator
+    val requestingParties = request.parties.toSet.map(Ref.Party.assertFromString)
+    val requestedApplicationId: domain.ApplicationId =
+      domain.ApplicationId(Ref.LedgerString.assertFromString(request.applicationId))
+
+    completionsService
+      .getCompletions(requestedOffset, requestedApplicationId, requestingParties)
+      .map { ce =>
+        val checkpoint = Some(
+          Checkpoint(
+            Some(fromInstant(ce.recordTime)),
+            Some(LedgerOffset(LedgerOffset.Value.Absolute(ce.offset.value)))))
+
+        ce match {
+          case CompletionEvent.CommandAccepted(_, _, commandId, transactionId) =>
+            //TODO: code smell, if we don't send anything for empty command IDs, why do we get the event at all?
```

Comment:
what do you think @gerolf-da ?

Reply:
The completion service should only emit events for "local" transactions.